sympy/functions/elementary/integers.py

```diff
@@ -144,8 +144,10 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
         arg = self.args[0]
         arg0 = arg.subs(x, 0)
         r = self.subs(x, 0)
-        if arg0.is_finite:
+        if arg0.is_finite or arg0 is S.NaN:
```

Comment:
These two cases should probably be treated in separation. If `arg0` is nan (i.e. subs fails), then limit of arg should be attempted.

Reply:
Cool , completely agreed . I had thought of going for that , but this seemed to be a viable option . I'll switch to this in the next commit !

Reply:
I've made a commit fixing this , this looks better than the previous commit to me , what do you think here ?

Reply:
Actually I realize , we won't be able to handle bi-directional limits through the latest commit
```
        if arg0 is S.NaN:
            limit = arg.limit(x, 0, dir='-' if re(cdir).is_negative else '+')
            ndir = arg.dir(x, cdir=cdir)
            return limit if ndir.is_negative else limit + 1
```
Consider calculating `limit(ceiling(sin(x)/x), x, 0, '+-')` , now through this we clearly get limits 1 from both sides but we won't be able to work this for the bi-directional case
```
>>> x = Symbol('x')
>>> limit(ceiling(sin(x)/x), x, 0)
1
>>> limit(ceiling(sin(x)/x), x, 0, '-')
1
>>> limit(ceiling(sin(x)/x), x, 0, '+-')
2
```
This is because through 
```
# line 1 limit = arg.limit(x, 0, dir='-' if re(cdir).is_negative else '+') we get limit as 1
# line 2 ndir = arg.dir(x, cdir=cdir) we get ndir as 0
# line 3 return limit if ndir.is_negative else limit + 1 we get answer as 2
```
To fix this we need a similar logic to that as the one written for cases where `arg0 is finite`
```
        if arg0.is_finite:
            if arg0 == r:
                if cdir == 0:
                    ndirl = arg.dir(x, cdir=-1)
                    ndir = arg.dir(x, cdir=1)
                    if ndir != ndirl:
                        raise ValueError("Two sided limit of %s around 0"
                                    "does not exist" % self)
                else:
                    ndir = arg.dir(x, cdir=cdir)
```
And clearly there is no point writing this twice !!! which could be done here . Hence I'll make the change for this , you could let me know if that could be improved .
Otherwise we won't be correctly solving the issue which states (**Bidirectional limit of floor(sin(x)/x) at 0, is not correct**) as of now . I'll also add relevant tests for this !

Reply:
I am quite happy with the latest commit here , answers all cases correctly . I guess this might be good to go once you review this !

Reply:
arg0 should not be NaN in this block. limit should be taken before this line.

Reply:
> arg0 should not be NaN in this block. limit should be taken before this line.
Ohh so you mean the second commit was correct ?
```
-        if arg0.is_finite or arg0 is S.NaN:
+       if arg0 is S.NaN:
+           limit = arg.limit(x, 0, dir='-' if re(cdir).is_negative else '+')
+           ndir = arg.dir(x, cdir=cdir)
+            return limit - 1 if ndir.is_negative else limit
        if arg0.is_finite:
            if arg0 == r:
                if cdir == 0:
                    ndirl = arg.dir(x, cdir=-1)
                    ndir = arg.dir(x, cdir=1)
```

Reply:
okay now that you've pointed that there should be a separate block, I would like to know your thoughts on this . I feel this would be the best way to go here . Because if we don't go for the same block there will be duplication of the following check shown below , which wouldn't be the correct way to go about here I presume.
```
        if cdir == 0:
            ndirl = arg.dir(x, cdir=-1)
            ndir = arg.dir(x, cdir=1)
            if ndir != ndirl:
                raise ValueError("Two sided limit of %s around 0"
                                 "does not exist" % self)
```
Hence we could go with this and use the check for the bi-directional cases of both (when **arg0 is finite or is S.NaN**)
```
    def _eval_as_leading_term(self, x, logx=None, cdir=0):
        arg = self.args[0]
        arg0 = arg.subs(x, 0)
        r = self.subs(x, 0)
        if cdir == 0:
            ndirl = arg.dir(x, cdir=-1)
            ndir = arg.dir(x, cdir=1)
            if ndir != ndirl:
                raise ValueError("Two sided limit of %s around 0"
                                 "does not exist" % self)
        else:
            ndir = arg.dir(x, cdir=cdir)
        if arg0 is S.NaN:
            limit = arg.limit(x, 0, dir='-' if re(cdir).is_negative else '+')
            return limit if ndir.is_negative else limit + 1
        if arg0.is_finite:
            if arg0 == r:
                return r if ndir.is_negative else r + 1
            else:
                return r
        return arg.as_leading_term(x, logx=logx, cdir=cdir)
```

Reply:
@jksuom , I've reverted the last commit , as the second commit would be according to what  you would prefer and maybe like to have !!